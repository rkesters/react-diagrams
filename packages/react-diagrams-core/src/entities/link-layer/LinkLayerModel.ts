import { LayerModel, LayerModelGenerics } from '@projectstorm/react-canvas-core';
import { LinkModel } from '../link/LinkModel';
import { DiagramEngine } from '../../DiagramEngine';
import { DiagramModel } from '../../models/DiagramModel';
import debug from 'debug';

const dbg = debug('reactDiagrams:diagramCore:LinkLayerModel');
export interface LinkLayerModelGenerics extends LayerModelGenerics {
	CHILDREN: LinkModel;
	ENGINE: DiagramEngine;
}

export class LinkLayerModel<G extends LinkLayerModelGenerics = LinkLayerModelGenerics> extends LayerModel<G> {
	constructor() {
		super({
			type: 'diagram-links',
			isSvg: true,
			transformed: true
		});
	}

	addModel(model: G['CHILDREN']): void {
		if (!(model instanceof LinkModel)) {
			throw new Error('Can only add links to this layer');
		}
		dbg(`Calling registerListener on ${this.getID()}`);
		model.registerListener({
			entityRemoved: () => {
				dbg(`executing entityRemoved handler`);
				(this.getParent() as DiagramModel).removeLink(model);
			}
		});
		super.addModel(model);
	}

	getLinks() {
		return this.getModels();
	}

	getChildModelFactoryBank(engine: G['ENGINE']) {
		return engine.getLinkFactories();
	}
}
