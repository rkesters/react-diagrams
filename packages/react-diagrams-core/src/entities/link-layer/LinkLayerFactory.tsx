import * as React from 'react';
import { AbstractReactFactory, GenerateModelEvent, GenerateWidgetEvent } from '@projectstorm/react-canvas-core';
import { DiagramEngine } from '../../DiagramEngine';
import { LinkLayerModel } from './LinkLayerModel';
import { LinkLayerWidget } from './LinkLayerWidget';

export class LinkLayerFactory extends AbstractReactFactory<LinkLayerModel, DiagramEngine> {
	constructor(engine: DiagramEngine) {
		super('diagram-links', engine);
	}

	generateModel(event: GenerateModelEvent): LinkLayerModel {
		return new LinkLayerModel();
	}

	generateReactWidget(event: GenerateWidgetEvent<LinkLayerModel>): JSX.Element {
		return <LinkLayerWidget layer={event.model} engine={this.engine} />;
	}
}
