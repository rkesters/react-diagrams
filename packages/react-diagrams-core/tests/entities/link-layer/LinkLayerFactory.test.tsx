import { LinkLayerFactory } from '../../../src/entities/link-layer/LinkLayerFactory';
import { LinkLayerModel } from '../../../src/entities/link-layer/LinkLayerModel';
import { render, waitFor } from '@testing-library/react';
import React from 'react';

describe('LinkLayerFactory', () => {
	test('generateModel', () => {
		const factory = new LinkLayerFactory();

		const model = factory.generateModel({});

		expect(model).toBeInstanceOf(LinkLayerModel);
	});

	test('generateReactWidget', () => {
		const factory = new LinkLayerFactory();

		const model = factory.generateModel({});

		const widget = render(factory.generateReactWidget({ model }), { wrapper: ({ children }) => <svg>{children}</svg> });

		expect(widget.baseElement).toMatchSnapshot();
	});
});
