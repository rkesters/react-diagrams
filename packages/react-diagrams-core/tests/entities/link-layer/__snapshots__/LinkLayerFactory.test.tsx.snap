// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkLayerFactory generateReactWidget 1`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;
